### PR TITLE
contour: 0.3.1.200 -> 0.3.12.262

### PR DESCRIPTION
**Pull request by moni-dz**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by fortuneteller2k** [comment]
I will fix the editorconfig complaint later.

**Comment by christianparpart** [comment]
> darwin failed with
> 
> ```
> In file included from /tmp/nix-build-contour-0.3.12.262.drv-0/source/build/src/contour/display/ContourTerminalDisplay_autogen/mocs_compilation.cpp:2:
> In file included from /tmp/nix-build-contour-0.3.12.262.drv-0/source/build/src/contour/display/ContourTerminalDisplay_autogen/EWIEGA46WW/moc_TerminalWidget.cpp:9:
> In file included from /tmp/nix-build-contour-0.3.12.262.drv-0/source/build/src/contour/display/ContourTerminalDisplay_autogen/EWIEGA46WW/../../../../../../src/contour/displ>
> In file included from /tmp/nix-build-contour-0.3.12.262.drv-0/source/src/contour/TerminalSession.h:17:
> /tmp/nix-build-contour-0.3.12.262.drv-0/source/src/contour/Audio.h:11:14: fatal error: 'QtMultimedia/QAudioSink' file not found
>     #include <QtMultimedia/QAudioSink>
>              ^~~~~~~~~~~~~~~~~~~~~~~~~
> In file included from /tmp/nix-build-contour-0.3.12.262.drv-0/source/src/contour/display/TerminalWidget.cpp:16:
> In file included from /tmp/nix-build-contour-0.3.12.262.drv-0/source/src/contour/ContourGuiApp.h:18:
> In file included from /tmp/nix-build-contour-0.3.12.262.drv-0/source/src/contour/TerminalSessionManager.h:3:
> In file included from /tmp/nix-build-contour-0.3.12.262.drv-0/source/src/contour/TerminalSession.h:17:
> /tmp/nix-build-contour-0.3.12.262.drv-0/source/src/contour/Audio.h:11:14: fatal error: 'QtMultimedia/QAudioSink' file not found
>     #include <QtMultimedia/QAudioSink>
>              ^~~~~~~~~~~~~~~~~~~~~~~~~
> 1 error generated.
> ```
> 
> for some reason

this may be because you may not have the Qt multimedia package (dev headers) installed on OS/X. I looked at the diff, it seems like you default to Qt6. If that turns out to be non-trivial on OS/X, I'd generally recommend to stick with Qt5 for the time being (just set the passed var to 5 instead of 6 and update the dependencies respectively).

**Comment by fortuneteller2k** [comment]
patching `src/contour/display/CMakeLists.txt` to include `Multimedia` and `Qt6::Multimedia` did the trick

**Comment by fortuneteller2k** [comment]
Result of `nixpkgs-review pr 238711` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>contour</li>
    <li>libunicode</li>
    <li>termbench-pro</li>
  </ul>
</details>

**Comment by fortuneteller2k** [comment]
I don't own an aarch64-linux machine (I don't have time to setup NixOS on Apple Silicon), so I'll wait on ofborg to test aarch64-linux.

**Comment by fortuneteller2k** [comment]
Result of `nixpkgs-review pr 238711` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>contour</li>
    <li>libunicode</li>
    <li>termbench-pro</li>
  </ul>
</details>

**Comment by christianparpart** [comment]
Many thanks for your work, @fortuneteller2k.
Is there any reason why the checkbox on "aarch64-linux" is not ticked above? (that should be working right out of the box as well)

**Comment by fortuneteller2k** [comment]
> Many thanks for your work, @fortuneteller2k. Is there any reason why the checkbox on "aarch64-linux" is not ticked above? (that should be working right out of the box as well)

Mainly because I haven't built it with Nix at the time, but building it on `aarch64-linux` but now I have:

```
moni ♥ nix build .#contour --system aarch64-linux --max-jobs 0                                                                                              nixpkgs → contour

moni ♥ readlink result                                                                                                                                      nixpkgs → contour
/nix/store/k4v6x5z0k2345s38pv7nnfi3cbzbbbsh-contour-0.3.12.262

moni ♥ file result/bin/contour                                                                                                                              nixpkgs → contour
result/bin/contour: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /nix/store/n8hgw8lja68x3zhv2ww9a7b6npbkvh21-glibc-2.37-8/lib/ld-linux-aarch64.so.1, for GNU/Linux 3.10.0, not stripped
```

File: pkgs/applications/terminal-emulators/contour/contour-cmakelists.diff [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
++   find_package(Qt6 COMPONENTS Core Gui OpenGLWidgets Widgets Multimedia REQUIRED)
+ else()
+    find_package(Qt5 COMPONENTS Gui Widgets REQUIRED)  # apt install qtbase5-dev libqt5gui5
+ endif()
+@@ -31,7 +31,7 @@ endif()
+ target_include_directories(ContourTerminalDisplay PRIVATE "${CMAKE_CURRENT_BINARY_DIR}/../..")
+ target_link_libraries(ContourTerminalDisplay vtrasterizer)
+ if(CONTOUR_QT_VERSION EQUAL "6")
+-    target_link_libraries(ContourTerminalDisplay Qt6::Core Qt6::Gui Qt6::OpenGL Qt6::OpenGLWidgets Qt6::Widgets)
++    target_link_libraries(ContourTerminalDisplay Qt6::Core Qt6::Gui Qt6::OpenGL Qt6::OpenGLWidgets Qt6::Widgets Qt6::Multimedia)
```
Comment (zopieux) [inline pkgs/applications/terminal-emulators/contour/contour-cmakelists.diff:17]: Could the missing `Qt6::Multimedia` dependency patch be sent upstream if it's actually required?

Comment (fortuneteller2k) [reply pkgs/applications/terminal-emulators/contour/contour-cmakelists.diff:17]: Wouldn't hurt, it's only required on Darwin for some reason.

**Review by vifino — APPROVED** [review]
This looks good to me.

Not sure about the separate QT6 commit though, maybe squash that with the previous one?

**Comment by christianparpart** [comment]
Hey guys. Many thanks for your support of inclusion into nixpkgs.
I just wanted to drop the little heads-up, that Qt6 will be the recommended default for future releases of Contour, too. If you find anything that is not nicely done, we're very gladly taking your patches upstream. :)

**Comment by moni-dz** [comment]
Darwin should be all good now, thanks to the recently landed stdenv rework, and codesigning the bundle executable so it doesn't segfault.

**Comment by moni-dz** [comment]
Result of `nixpkgs-review pr 238711` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>contour</li>
    <li>contour.terminfo</li>
  </ul>
</details>

**Comment by moni-dz** [comment]
Result of `nixpkgs-review pr 238711` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>contour</li>
    <li>contour.terminfo</li>
  </ul>
</details>

File: pkgs/applications/terminal-emulators/contour/default.nix [hunk 2]
```diff
@@ -16,94 +20,87 @@
...
+    ln -s $out/bin $out/Applications/contour.app/Contents/MacOS
+  '' + lib.optionalString stdenv.isLinux ''
     mv $out/share/terminfo $terminfo/share/
+    installShellCompletion --zsh $out/share/contour/shell-integration/shell-integration.zsh
+    installShellCompletion --fish $out/share/contour/shell-integration/shell-integration.fish
+  '' + ''
     echo "$terminfo" >> $out/nix-support/propagated-user-env-packages
   '';
 
   passthru.tests.test = nixosTests.terminal-emulators.contour;
```
Comment (pbsds) [inline pkgs/applications/terminal-emulators/contour/default.nix:96]:
```
error: attribute 'contour' missing

       at /dev/shm/nixpkgs-review/pr-238711/nixpkgs/pkgs/applications/terminal-emulators/contour/default.nix:96:25:

           95|
           96|   passthru.tests.test = nixosTests.terminal-emulators.contour;
             |                         ^
           97|
```

It seems `nixos/tests/terminal-emulators.nix` is missing an entry.

File: pkgs/development/libraries/libunicode/default.nix [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    rev = "65e0c6ddf9648b94aa9bc7dda0718401efa9ef8e";
+    hash = "sha256-F4CVU5MImkM571mD4iFxqTnNbk2GXKTGksqO4LH2uEk=";
+  };
+
+  nativeBuildInputs = [ cmake python3 ];
+  buildInputs = [ catch2 fmt ];
+
+  cmakeFlags = [ "-DLIBUNICODE_UCD_DIR=${ucd-src}" ];
+
+  meta = with lib; {
```
Comment (pbsds) [inline pkgs/development/libraries/libunicode/default.nix:27]: please set `platforms`

File: pkgs/development/libraries/termbench-pro/default.nix [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+    mkdir -p $out/bin
+    mkdir -p $out/lib
+    mv termbenchpro/tbp $out/bin
+    mv libtermbench/libtermbench.a $out/lib
+
+    runHook postInstall
+  '';
+
+  meta = with lib; {
```
Comment (pbsds) [inline pkgs/development/libraries/termbench-pro/default.nix:28]: Please set `platforms`

**Review by pbsds — COMMENTED** [review]
Superseeds and closes #271160

Result of `nixpkgs-review pr 238711` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>contour</li>
    <li>contour.terminfo</li>
    <li>libunicode</li>
    <li>termbench-pro</li>
  </ul>
</details>


- [x] package path fits guidelines
- [x] package name fits guidelines
- [ ] package version fits guidelines
- [x] package builds on x86_64-linux
- [x] executables tested on x86_64-linux
- [x] `meta.description` is set and fits guidelines
- [x] `meta.license` fits upstream license
- [ ] `meta.platforms` is set
- [x] `meta.maintainers` is set
- [x] build time only dependencies are declared in `nativeBuildInputs`
- [x] source is fetched using the appropriate function
- [x] the list of `phases` is not overridden
- [x] when a phase (like `installPhase`) is overridden it starts with `runHook preInstall` and ends with `runHook postInstall`.
- [x] patches have a comment describing either the upstream URL or a reason why the patch wasn't upstreamed
- [x] patches that are remotely available are fetched rather than vendored

File: pkgs/development/libraries/termbench-pro/default.nix [hunk 5]
```diff
@@ -0,0 +1,33 @@
+{ lib, stdenv, fetchFromGitHub, cmake, fmt }:
+
+stdenv.mkDerivation {
+  pname = "termbench-pro";
+  version = "unstable-2023-01-26";
```
[the remark this answers is not in the document]
Comment (vifino) [reply pkgs/development/libraries/termbench-pro/default.nix:5]: I think the *should* should be ignored in this case - version 0.0.0 adds no value.

File: pkgs/development/libraries/libunicode/default.nix [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  ucd-version = "15.0.0";
+
+  ucd-src = fetchzip {
+    url = "https://www.unicode.org/Public/${ucd-version}/ucd/UCD.zip";
+    hash = "sha256-jj6bX46VcnH7vpc9GwM9gArG+hSPbOGL6E4SaVd0s60=";
+    stripRoot = false;
+  };
+in stdenv.mkDerivation (final: {
+  pname = "libunicode";
+  version = "unstable-2023-03-05";
```
[the remark this answers is not in the document]
Comment (christianparpart) [reply pkgs/development/libraries/libunicode/default.nix:13]: (old text start)
um no, not yet? it's about to be released ASAP, but I don't want to give any day-count for that. We want to hunt down some more bugs. It's okay to start with 0.3.12 still, unless you guys (nix packaging team) are okay with pre-release snapshots, then we could do a 0.4.0rc1 today. It's about 2 or 3 bugs that we want to get fixed before release.
(old text end)

EDIT: I just realized this is about `libunicode`, No, please stick with the version you are having. the libunicode 0.4.0 release is specificially for Contour 0.4.0 (to be released asap, as mentioned above). Contour 0.3.12 works best with the respective libunicode 0.3's version :) - sorry for the confusion) :)

File: pkgs/development/libraries/termbench-pro/default.nix (hunk 5 above)
[the remark this answers is not in the document]
Comment (christianparpart) [reply pkgs/development/libraries/termbench-pro/default.nix:5]: termbench-pro can be skipped for building (and packaging). This is a maintainer's dependency to investigate on potential performance bottlenecks. Thanks for bringing this up, we can indeed make it a release soonish (though, this is not top priority right now :) )

File: pkgs/development/libraries/libunicode/default.nix (hunk 3 above)
Comment (christianparpart) [reply pkgs/development/libraries/libunicode/default.nix:27]: NB: we test on x86-64 and on ARM64

**Comment by pbsds** [comment]
waiting for ofborg

**Comment by pbsds** [comment]
I think this fixes the test:
```diff
diff --git a/nixos/tests/terminal-emulators.nix b/nixos/tests/terminal-emulators.nix
index d7963866d4e2..e53bfa7453bc 100644
--- a/nixos/tests/terminal-emulators.nix
+++ b/nixos/tests/terminal-emulators.nix
@@ -24,7 +24,7 @@ let tests = {
       alacritty.pkg = p: p.alacritty;
 
       contour.pkg = p: p.contour;
-      contour.cmd = "contour $command";
+      contour.cmd = "contour early-exit-threshold 0 execute $command";
 
       cool-retro-term.pkg = p: p.cool-retro-term;
       cool-retro-term.colourTest = false; # broken by gloss effect
```

**Comment by moni-dz** [comment]
> Is it you I should ping for upcoming releases of Contour, or do you get notified already by Github? (0.4.0 should be released this year finally, too)

Pings would be fine, I don't usually watch repositories because it gets unruly in my inbox.

**Pull request merged**